src/treemodel/liverange.js

```diff
@@ -0,0 +1,139 @@
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+CKEDITOR.define( [
+	'treemodel/range',
+	'treemodel/liveposition',
+	'emittermixin',
+	'utils'
+], ( Range, LivePosition, EmitterMixin, utils ) => {
+	/**
+	 * LiveRange is a Range in the Tree Model that updates itself as the tree changes. It may be used as a bookmark.
+	 * **Note:** Be very careful when dealing with LiveRange. Each LiveRange instance bind events that might
+	 * have to be unbound. Use {@link #detach} whenever you don't need LiveRange anymore.
+	 *
+	 * @class treeModel.LiveRange
+	 */
+	class LiveRange extends Range {
+		/**
+		 * Creates a live range.
+		 *
+		 * @see {treeModel.Range}
+		 * @constructor
+		 */
+		constructor( start, end ) {
+			super( start, end );
+
+			this.start = new LivePosition( this.start.root, this.start.path.slice(), LivePosition.STICKS_TO_NEXT );
+			this.end = new LivePosition( this.end.root, this.end.path.slice(), LivePosition.STICKS_TO_PREVIOUS );
+
+			bindWithDocument.call( this );
+		}
+
+		/**
+		 * Unbinds all events previously bound by LiveRange. Use it whenever you don't need LiveRange instance
+		 * anymore (i.e. when leaving scope in which it was declared or before re-assigning variable that was
+		 * referring to it).
+		 */
+		detach() {
+			this.start.detach();
+			this.end.detach();
+			this.stopListening();
+		}
+
+		/**
+		 * @see {@link treeModel.Range#createFromElement}
+		 * @static
+		 * @method createFromElement
+		 * @param {treeModel.Element} element
+		 * @returns {treeModel.LiveRange}
+		 */
+
+		/**
+		 * @see {@link treeModel.Range#createFromPositionAndShift}
+		 * @static
+		 * @method createFromPositionAndShift
+		 * @param {treeModel.Position} position
+		 * @param {Number} shift
+		 * @returns {treeModel.LiveRange}
+		 */
+
+		/**
+		 * @see {@link treeModel.Range#createFromParentsAndOffsets}
+		 * @static
+		 * @method createFromParentsAndOffsets
+		 * @param {treeModel.Element} startElement
+		 * @param {Number} startOffset
+		 * @param {treeModel.Element} endElement
+		 * @param {Number} endOffset
+		 * @returns {treeModel.LiveRange}
+		 */
+
+		/**
+		 * @see {@link treeModel.Range#createFromRange}
+		 * @static
+		 * @method createFromRange
+		 * @param {treeModel.Range} range
+		 * @returns {treeModel.LiveRange}
+		 */
+	}
+
+	/**
+	 * Binds this LiveRange to the {@link treeModel.Document} that owns this range.
+	 *
+	 * @private
+	 * @method bindWithDocument
+	 */
+	function bindWithDocument() {
+		/*jshint validthis: true */
+
+		this.listenTo(
+			this.root.document,
+			'change',
+			( event, type, changes ) => {
+				fixBoundaries.call( this, type, changes.range, changes.sourcePosition );
+			},
+			this
+		);
+	}
+
+	/**
+	 * LiveRange is partially updated by "automatic" updates of LivePositions that are boundaries of LiveRange.
+	 * However there are cases when the boundaries have to be fixed because they end up in wrong places.
```

Comment:
"there are cases" What cases? Example would be useful. You can put such example in the class description to show what is the difference between `liveRange` and range of `livePosition`s.